app/models/miq_queue.rb

```diff
@@ -322,31 +322,27 @@ def self.peek(options = {})
   #   used to put a new item on the queue.
   #
   def self.put_or_update(find_options)
-    find_options  = default_get_options(find_options)
-    conds = find_options.dup
+    find_options = default_get_options(find_options)
 
     # Since args are a serializable field, remove them and manually dump them
-    #   for proper comparison.  NOTE: hashes may not compare correctly due to
-    #   it's unordered nature.
-    where_scope = if conds.key?(:args)
-                    args = YAML.dump(conds.delete(:args))
-                    MiqQueue.where(conds).where(['args = ?', args])
-                  else
-                    MiqQueue.where(conds)
-                  end
+    #   for proper comparison.
+    where_scope =
+      if find_options.key?(:args)
+        MiqQueue.where(find_options.except(:args)).where(['args = ?', find_options[:args].try(:to_yaml)])
+      else
+        MiqQueue.where(find_options)
+      end
 
     msg = nil
     loop do
       msg = where_scope.order("priority, id").first
 
-      save_options = block_given? ? yield(msg, find_options) : nil
-      save_options = save_options.dup unless save_options.nil?
+      save_options = block_given? ? yield(msg, find_options.dup) : nil
```

Comment:
I wish I could get rid of this .dup, but the block modify the caller, and I would have to change each one individually :(

Reply:
Oh wait...the original code passed the *ORIGINAL* find_options, which means I *can* get rid of this dup, I think.